lisp/ess-r-mode.el

```diff
@@ -2352,7 +2352,8 @@ If the current buffer does not have a usage section, return nil."
           (goto-char (point-min))
           (forward-whitespace 1)
           (while (not (eobp))
-            (if (looking-at (rx (group (1+ (not (any whitespace)))) "("))
+            (if (looking-at "#") (forward-line 1))
```

Comment:
What if there are multiple `#` lines? Or spaces in front of `#`? I think this will turn out fine as iterations progress because of the forward-space below but the flow is not very clear. Moving this just after the forward-space would be clearer.

(Ideally you would use `(ess-skip-blanks-forward t)` which skips blanks, newlines and comments instead of the `forward-space`. However it won't work here because in help buffers the syntax class of `#` is punctuation instead of comment opener.)

Style: When there is no else clause we use a `when` instead of an `if`. Also the then clause should generally be on its own line, especially if it performs a side effect such as moving cursor.

Reply:
Style: got it, thank you.

Multiple `#` lines are all cleaned individually. Spaces in front of the `#` are skipped by the `(forward-whitespace 1)`. I had looked into the use of comment-skipping elisp as well, but as you said, `#` is not a comment in that context.

I think you're saying you'd recommend

```elisp
          (while (not (eobp))
            (if (looking-at (rx (group (1+ (not (any "#" whitespace)))) "("))
                (progn
                  (push (match-string-no-properties 1) usage-objects)
                  ;; Skip past function arguments
                  (forward-list))
              (if (looking-at "#")
                  (forward-line 1)
                (forward-whitespace 1))))
```

or perhaps switch to `cond` with

```elisp
          (while (not (eobp))
            (cond ((looking-at (rx (group (1+ (not (any "#" whitespace)))) "("))
                   (progn
                     (push (match-string-no-properties 1) usage-objects)
                     ;; Skip past function arguments
                     (forward-list)))
                  ((looking-at "#") (forward-line 1))
                  (t (forward-whitespace 1))))
```

(This point is moot if you think an R implementation is safer/faster/better.)

Side question: is there a particular reason there is raw regex in the `when-let` but the use of `rx` for generating a regex that is no more complex? I understand some prefer the readability of `rx`-generated expressions, but I'd think *these* patterns would be trivial-enough to maintain. The only other use of `rx` (flymake) does make a non-trivial regex, but there are plenty of other non-`rx` patterns (some rather challenging) throughout the code.

---

<details><summary>Sample tests (manually run)</summary>

```
read.table                package:utils                R Documentation

Data Input

Description:

     Reads a file in table format and creates a data frame from it,
     with cases corresponding to lines and variables to fields in the
     file.

Usage:

     read.table(file, header = FALSE, sep = "", quote = "\"'",
                dec = ".", numerals = c("allow.loss", "warn.loss", "no.loss"),
                row.names, col.names, as.is = !stringsAsFactors,
                na.strings = "NA", colClasses = NA, nrows = -1,
                skip = 0, check.names = TRUE, fill = !blank.lines.skip,
                strip.white = FALSE, blank.lines.skip = TRUE,
                comment.char = "#",
                allowEscapes = FALSE, flush = FALSE,
                stringsAsFactors = default.stringsAsFactors(), # something
                fileEncoding = "", encoding = "unknown", text, skipNul = FALSE)
     
     # the main function everybody uses
     read.csv(file, header = TRUE, sep = ",", quote = "\"",
              dec = ".", fill = TRUE, comment.char = "", ...)
     
     # ... same, but everywhere except US
     read.csv2(file, header = TRUE, sep = ";", quote = "\"",
               dec = ",", fill = TRUE, comment.char = "", ...)
     
     # weird data
     read.delim(file, header = TRUE, sep = "\t", quote = "\"",
                dec = ".", fill = TRUE, comment.char = "", ...)
     
     read.delim2(file, header = TRUE, sep = "\t", quote = "\"",
                 dec = ",", fill = TRUE, comment.char = "", ...)
     
Arguments:
```

returns

```elisp
(("read.table" "file" "header" "sep" "quote" "dec" "numerals" "row.names" "col.names" "as.is" "na.strings" "colClasses" ...) ("read.csv" "file" "header" "sep" "quote" "dec" "fill" "comment.char" "...") ("read.csv2" "file" "header" "sep" "quote" "dec" "fill" "comment.char" "...") ("read.delim" "file" "header" "sep" "quote" "dec" "fill" "comment.char" "...") ("read.delim2" "file" "header" "sep" "quote" "dec" "fill" "comment.char" "..."))
```

---

```
J                  package:data.table                  R Documentation

Creates a join 'data.table'

Description:

     Creates a 'data.table' for use in 'i' in a '[.data.table' join.

Usage:

     # DT[J(...)]                          # J() only for use inside DT[...]
     # DT[.(...)]                          # .() only for use inside DT[...]
     # DT[list(...)]                       # same; .(), list() and J() are identical
     SJ(...)                             # DT[SJ(...)]
     CJ(..., sorted=TRUE, unique=FALSE)  # DT[CJ(...)]
     
Arguments:
```

returns

```
(("SJ" "...") ("CJ" "..." "sorted" "unique"))
```

---

Same help-page, but with some blank-space changes around comment-chars to check some oddities:

```
J                  package:data.table                  R Documentation

Creates a join 'data.table'

Description:

     Creates a 'data.table' for use in 'i' in a '[.data.table' join.

Usage:

     ## DT[J(...)]                          # J() only for use inside DT[...]
     # # DT[.(...)]                          # .() only for use inside DT[...]
     #DT[list(...)]                       # same; .(), list() and J() are identical
     SJ(...)#DT[SJ(...)]
     CJ(..., sorted=TRUE, unique=FALSE)  # DT[CJ(...)]
     
Arguments:
```

also returns

```
(("SJ" "...") ("CJ" "..." "sorted" "unique"))
```

</details>

Reply:
Regarding the use of `rx`, I don't know. I had to run it locally to see the generated regexp since I'm not familiar with this syntax.

I like the `cond`, it's quite readable and clear. There should be newlines after the conditions though (including `t`). Also the `progn` can be removed since all branches are implicitly progns:

```lisp
(while (not (eobp))
  (cond ((looking-at (rx (group (1+ (not (any "#" whitespace)))) "("))
         (push (match-string-no-properties 1) usage-objects)
         ;; Skip past function arguments
         (forward-list))
        ((looking-at "#")
         (forward-line 1))
        (t
         (forward-whitespace 1))))
```

This way it's easier to see what is a condition and what is a branch.